src/Components/MapModal.js

```diff
@@ -110,9 +116,83 @@ export default class ExampleApp extends React.Component {
     handleCloseModal () {
         this.props.hideModal();
     }
+
+    incrementPagination(numPages, currentPage) {
+        if (currentPage < numPages) {
+            this.setState((state, props) => ({
+                currentPage: state.currentPage+1
+            }));
+        }
+    }
+
+    decrementPagination(numPages, currentPage) {
+        if (currentPage > 1) {
+            this.setState((state, props) => ({
+                currentPage: state.currentPage-1
+            }));
+        }
+    }
+
+    goToPage(pageNum) {
+        this.setState({currentPage: pageNum});
+    }
+
+    createPaginationItem(number, currentPage) {
+        return (
+            <Pagination.Item key={number} active={number === currentPage} onClick={() => this.goToPage(number)}>
+                {number}
+            </Pagination.Item>
+        );
+    }
+
+    createPagination(numPages, currentPage) {
+        let items = [];
+
+        items.push(<Pagination.Prev aria-label="Previous page" key={"prev"} onClick={() => this.decrementPagination(numPages, currentPage)} />);
+        if (currentPage > 3) { items.push(<Pagination.Ellipsis aria-label="More pages" key={"firstEllipsis"} />); }
+
+        if (currentPage < 3) {
+            for (let number = 1; number <= 5; number++) {
+                items.push(
+                    this.createPaginationItem(number, currentPage)
+                );
+            }
+        }
+        else if (currentPage > numPages-2) {
+            for (let number = numPages-4; number <= numPages; number++) {
+                items.push(
+                    this.createPaginationItem(number, currentPage)
+                );
+            }
+        }
+        else {
+            for (let number = currentPage-2; number <= currentPage+2; number++) {
+                items.push(
+                    this.createPaginationItem(number, currentPage)
+                );
+            }
+        }
+
+        if (currentPage < numPages-2) { items.push(<Pagination.Ellipsis aria-label="More pages" key={"secondEllipsis"} />); }
+        items.push(<Pagination.Next aria-label="Next page" key={"next"} onClick={() => this.incrementPagination(numPages, currentPage)} />);
+
+        const pagination = (
+            <div>
+                <Pagination>{items}</Pagination>
+            </div>
+        );
+        return pagination;
+    }
     
     render () {
-        const {sortedProgramResults} = this.state;
+        const {sortedProgramResults, currentPage, pageSize} = this.state;
+
+        let numPages = Math.ceil(sortedProgramResults.length / pageSize);
+
+        const shownResults = sortedProgramResults.slice((currentPage-1)*pageSize, (currentPage-1)*pageSize + 5);
+
+        const pagination = this.createPagination(numPages, currentPage);
```

Comment:
for optimization, we should do this only when the component loads or when a page is changed. render() can be run for many reasons, and we wouldn't want to compute all the pagination stuff again when it didn't change!